pkg/telemetry/otel.go

```diff
@@ -65,11 +79,41 @@ func InstrumentLambdaHandler(handlerFunc interface{}) interface{} {
 
 	return otellambda.InstrumentHandler(
 		handlerFunc,
+		otellambda.WithEventToCarrier(jsonEventHeadersToCarrier),
```

Comment:
Interesting: https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-lambda-go/otellambda#EventToCarrier
> An EventToCarrier function defines how the instrumentation should prepare a TextMapCarrier for the configured propagator to read from. This extra step is necessary because Lambda does not have HTTP headers to read from and instead stores the headers it was invoked with (including TraceID, etc.) as part of the invocation event. If using the AWS XRay tracing then the trace information is instead stored in the Lambda environment.

So this is technically the "fix" for the issue, great find!

Reply:
yeah. It all comes from the fact that lambda handlers can have different shapes and receive many event schemas. Honestly, I would've expected `otellambda.InstrumentHandler` to take care of most of that automatically, at least for the most common events, such as the ones produced by the API Gateway.

Surprisingly enough, the headers are correctly transferred to the request injected into the lambda handler for HTTP requests, so using the standard OpenTelemetry libs for `net/http` would've been an alternative solution and would've worked out of the box. This approach, however, gives us a bit more lambda-specific information.